eland/ml/pytorch/external_models.py

```diff
@@ -0,0 +1,391 @@
+
+            return torch.jit.trace(self._model, (tokens_tensor, attention_mask, token_type_ids, position_ids))
+
+    class TextClassificationModel(TraceableClassificationModel):
+        def trace(self) -> TracedModelTypes:
+            # model needs to be in evaluate mode
+            self._model.eval()
+
+            # tokenizing dummy input text
+            text = "The cat was sick on the bed."
+            tokenized_text = self._tokenizer.tokenize(text)
+
+            # create token IDs
+            indexed_tokens = self._tokenizer.convert_tokens_to_ids(tokenized_text)
+
+            # prepare dummy input tensors
+            tokens_tensor = torch.tensor([indexed_tokens])
+            attention_mask = torch.ones(tokens_tensor.size())
+            token_type_ids = torch.zeros(tokens_tensor.size(), dtype=torch.long)
+            position_ids = torch.arange(tokens_tensor.size(1), dtype=torch.long)
+
+            return torch.jit.trace(self._model, (tokens_tensor, attention_mask, token_type_ids, position_ids))
+
+    class TextEmbeddingModel(TraceableModel):
+        def trace(self) -> TracedModelTypes:
+            # model needs to be in evaluate mode
+            self._model.eval()
+
+            # tokenizing dummy input text
+            text = "This is an example sentence"
+            tokenized_text = self._tokenizer.tokenize(text)
+
+            # create token IDs
+            indexed_tokens = self._tokenizer.convert_tokens_to_ids(tokenized_text)
+
+            # prepare dummy input tensors
+            tokens_tensor = torch.tensor([indexed_tokens])
+            attention_mask = torch.ones(tokens_tensor.size())
+            token_type_ids = torch.zeros(tokens_tensor.size(), dtype=torch.long)
+            position_ids = torch.arange(tokens_tensor.size(1), dtype=torch.long)
+
+            return torch.jit.trace(self._model, (tokens_tensor, attention_mask, token_type_ids, position_ids))
+
+    def __init__(self, model_id: str, task_type: str):
+        self._model_id = model_id
+        self._task_type = task_type.replace('-', '_')
+
+        # Elasticsearch model IDs need to be a specific format: no special chars, all lowercase, max 64 chars
+        self._es_model_id = self._model_id.replace('/', '__').lower()[:64]
+
+        # load Hugging Face model and tokenizer
+        self._tokenizer = transformers.AutoTokenizer.from_pretrained(self._model_id)
+
+        # check for a supported tokenizer
+        if not isinstance(self._tokenizer, SUPPORTED_TOKENIZERS):
+            raise TypeError(f"Tokenizer type {self._tokenizer} not supported, must be one of: {SUPPORTED_TOKENIZERS_STR}")
+
+        self._traceable_model = self._create_traceable_model()
+        self._traced_model = self._traceable_model.trace()
+        self._vocab = self._load_vocab()
+        self._config = self._create_config()
+
+    @staticmethod
+    def _dict_to_ordered_list(dict: Dict[Any, Any], sort_by_key: bool = False, sort_by_value: bool = False) -> List[Any]:
```

Comment:
Seems like a function instead of a staticmethod? Also since there are two distinct functionalities should probably be split into two?

```python
# sort_by_key
sorted_values = [v for _, v in sorted(dict.items(), key=lambda kv: kv[0])]

# sort_by_value
sorted_keys = [k for k, _ in sorted(dict.items(), key=lambda kv: kv[1])]
```

Reply:
I'm gonna do this change last, because I'm not sure it's possible. Let's come back to this.

Reply:
This is called from outside the class as well so I'm gonna leave it for now. Happy to review once this is on a branch, if there is a better way.